// starknet/dist/utils/uint256.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.bnToUint256 = exports.isUint256 = exports.UINT_256_MAX = exports.UINT_128_MAX = exports.uint256ToBN = void 0;
var encode_1 = require("./encode");
var number_1 = require("./number");
// function to convert Uint256 to BN
function uint256ToBN(uint256) {
    return (0, number_1.toBN)(uint256.high).shln(128).add((0, number_1.toBN)(uint256.low));
}
exports.uint256ToBN = uint256ToBN;
exports.UINT_128_MAX = (0, number_1.toBN)(1).shln(128).sub((0, number_1.toBN)(1));
exports.UINT_256_MAX = (0, number_1.toBN)(1).shln(256).sub((0, number_1.toBN)(1));
// function to check if BN is smaller or equal 2**256-1
function isUint256(bn) {
    return (0, number_1.toBN)(bn).lte(exports.UINT_256_MAX);
}
exports.isUint256 = isUint256;
// function to convert BN to Uint256
function bnToUint256(bignumber) {
    var bn = (0, number_1.toBN)(bignumber);
    if (!isUint256(bn))
        throw new Error('Number is too large');
    return {
        low: (0, encode_1.addHexPrefix)(bn.maskn(128).toString(16)),
        high: (0, encode_1.addHexPrefix)(bn.shrn(128).toString(16)),
    };
}
exports.bnToUint256 = bnToUint256;
