// @ledgerhq/cryptoassets/lib/abandonseed.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
exports.__esModule = true;
exports.getAbandonSeedAddress = void 0;
var invariant_1 = __importDefault(require("invariant"));
var abandonSeedAddresses = {
    algorand: "PSHLIWQKDEETIIBQEOTLGCT5IF7BTTOKCUULONOGVGF2HYDT2IHW3H4CCI",
    cosmos: "cosmos19rl4cm2hmr8afy4kldpxz3fka4jguq0auqdal4",
    cosmos_testnet: "cosmos19rl4cm2hmr8afy4kldpxz3fka4jguq0auqdal4",
    ripple: "rHsMGQEkVNJmpGWs8XUBoTBiAAbwxZN5v3",
    stellar: "GDYPMQMYW2JTLPWAUAHIDY3E4VHP5SGTFC5SMA45L7ZPOTHWQ2PHEW3E",
    tezos: "tz1VJitLYB31fEC82efFkLRU4AQUH9QgH3q6",
    tron: "0x0000000000000000000000000000000000000000",
    ethereum: "0x0000000000000000000000000000000000000000",
    bitcoin: "1LqBGSKuX5yYUonjxT5qGfpUsXKYYWeabA",
    bitcoin_cash: "1mW6fDEMjKrDHvLvoEsaeLxSCzZBf3Bfg",
    bitcoin_gold: "GeTZ7bjfXtGsyEcerSSFJNUSZwLfjtCJX9",
    bitcoin_private: "b1SGV7U5kGAMHtGbkAR3mjaZqVn57SHFbiR",
    bitcoin_testnet: "mkpZhYtJu2r87Js3pDiWJDmPte2NRZ8bJV",
    dash: "XoJA8qE3N2Y3jMLEtZ3vcN42qseZ8LvFf5",
    decred: "",
    // FIXME need to derivate decred abandon
    digibyte: "DG1KhhBKpsyWXTakHNezaDQ34focsXjN1i",
    dogecoin: "DBus3bamQjgJULBJtYXpEzDWQRwF5iwxgC",
    game_credits: "GJgbzWpGhrZmSvc2V5Npqf57Kg9xfB79tj",
    komodo: "RW8gfgpCUdgZbkPAs1uJQF2S9681JVkGRi",
    litecoin: "LUWPbpM43E2p7ZSh8cyTBEkvpHmr3cB8Ez",
    nix: "GRpn2DPiQxAczMrQFt2sK1CS8EYdnvSHxo",
    peercoin: "PFinP8Tm5hFJKfVnSMFfiNibPNfHgXzTDZ",
    pivx: "DDBxSas734KhMp1Btga3LdwWAc1igSER8o",
    polkadot: "111111111111111111111111111111111HC1",
    qtum: "QPvRe2C17qk24K6v5gTg7CPghZ8b4WMxZP",
    stakenet: "XhPU3ZrghEh387nqgQ1Fi16JvtkSeYdP4g",
    stealthcoin: "SKsLkKVeMtPNZQuNfUNXi3Bk1TwP1QPqJG",
    stratis: "Sdo6x9k5AxWtfyJe5B9SZPteYTKgUoMMr1",
    vertcoin: "Vce16eJifb7HpuoTFEBJyKNLsBJPo7fM83",
    viacoin: "VfwB3jbDWELpfiFNzbEVgx73HsR9bAq35C",
    zcash: "t1XVXWCvpMgBvUaed4XDqWtgQgJSu1Ghz7F",
    zclassic: "t1Qmwyih5F7Mw6Vts4tSnXuA2o3NgJPYNgP",
    zcoin: "a1bW3sVVUsLqgKuTMXtSaAHGvpxKwugxPH",
    zencash: "zngWJRgpBa45KUeRuCmdMsqti4ohhe9sVwC"
};
/**
 * Returns a valid address for a given currency.
 * These addresses are PUBLIC addresses
 * We use them for tests and also for dry-run estimations
 * DO NOT USE AS RECIPIENT OR SIGN TRANSACTIONS INTO THEM
 * @param {*} currencyId
 */
var getAbandonSeedAddress = function (currencyId) {
    (0, invariant_1["default"])(abandonSeedAddresses[currencyId] !== undefined, "No abandonseed available for ".concat(currencyId));
    return abandonSeedAddresses[currencyId];
};
exports.getAbandonSeedAddress = getAbandonSeedAddress;
